src/mlpack/methods/ann/layer/instance_norm_impl.hpp

```diff
@@ -0,0 +1,148 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_INSTANCE_NORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_INSTANCE_NORM_IMPL_HPP
+
+// In case it is not included.
+#include "instance_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm() :
+    size(0),
+    eps(1e-8),
+    average(true),
+    momentum(0.0),
+    deterministic(false),
+    reset(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm(
+    const size_t size,
+    const size_t batchSize,
+    const double eps,
+    const bool average,
+    const double momentum) :
+    size(size),
+    batchSize(batchSize),
+    eps(eps),
+    average(average),
+    momentum(momentum),
+    deterministic(false),
+    reset(false)
+{
+    batchNorm = ann::BatchNorm<> (size * batchSize,
+                                  eps,
+                                  average,
+                                  momentum);
+    runningMean.zeros(size, 1);
+    runningVariance.ones(size, 1);
+    runningVariance = batchNorm.TrainingVariance();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void InstanceNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
+{
+  // Instance Norm with (N, C, H, W) is same as Batch Norm with (1, N*C, H, W),
+  // where N is the batchSize, C is the number of channels, H and W are the
+  // height and width of each image respectively.
+  if (input.n_cols != batchSize)
+  {
+    Log::Fatal << "Must use the same BatchSize that was used in the constructor."
+        << std::endl;
+  }
+
+  if (!reset)
+  {
+    batchNorm.Reset();
+    reset = true;
+  }
+
+  shapeA = input.n_rows;
+  shapeB = input.n_cols;
+
+  if (deterministic)
+    batchNorm.Deterministic() = true;
+
+  arma::mat inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(),
+      shapeA * shapeB, 1, false, false);
+  batchNorm.Forward(inputTemp, output);
+  output.reshape(shapeA, shapeB);
+  runningMean = batchNorm.TrainingMean();
+  runningMean.reshape(size, shapeB);
+  runningMean = arma::mean(runningMean, 1);
+  runningVariance = batchNorm.TrainingVariance();
+  runningVariance.reshape(size, shapeB);
+  runningVariance = arma::mean(runningVariance, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void InstanceNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>& input,
+    const arma::Mat<eT>& gy,
+    arma::Mat<eT>& g)
+{
+  arma::mat inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(),
+      shapeA * shapeB, 1, false, false);
+  arma::mat gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(),
+      shapeA * shapeB, 1, false, false);
```

Comment:
Just curious, do we need to need store shapeA and shapeB (can't we calculate them from input?). I maybe missing something though.

Reply:
Yes, you are right, since input size in the forward pass = output size in the forward pass = input in the backward step is the output from the forward step.

Reply:
Agreed. Thanks for pointing it out :)